test/test_multitensor.py

```diff
@@ -69,6 +74,38 @@ def _test_simple_reduce_axis(self, shard_x):
     O = X.sum(axis=1)
     np.testing.assert_allclose(O.numpy(), 256)
 
+  def test_fuzz_allreduce(self):
```

Comment:
this is quite slow on CLANG backend, took 9 minutes to finish. not sure if it's expected

Reply:
That's not expected - will investigate why

Reply:
```
Program: /Users/user/src/tinygrad/.venv/bin/pyinstrument wtf.py

27.106 <module>  wtf.py:1
└─ 26.899 Tensor.numpy  tinygrad/tensor.py:143
   └─ 26.892 Tensor.realize  tinygrad/tensor.py:112
      └─ 26.767 run_schedule  tinygrad/realize.py:25
         └─ 26.587 lower_schedule_item  tinygrad/realize.py:16
            └─ 26.526 Compiled.get_runner  tinygrad/device.py:329
               ├─ 25.291 Compiled.to_program  tinygrad/device.py:292
               │  ├─ 23.976 CompiledASTRunner.build  tinygrad/device.py:261
               │  │  └─ 23.975 ClangProgram.__init__  tinygrad/runtime/ops_clang.py:16
               │  │     └─ 23.633 CDLL.__init__  ctypes/__init__.py:342
               │  │           [2 frames hidden]  ctypes, <built-in>
               │  │              23.620 dlopen  <built-in>
               │  └─ 1.096 Linearizer.linearize  tinygrad/codegen/linearizer.py:181
               │     └─ 0.432 <dictcomp>  tinygrad/codegen/linearizer.py:383
               │        └─ 0.412 Linearizer.global_load  tinygrad/codegen/linearizer.py:85
               └─ 1.224 Compiled.get_linearizer  tinygrad/device.py:305
                  ├─ 0.890 Linearizer.__init__  tinygrad/codegen/kernel.py:70
                  │  ├─ 0.436 get_lazyop_info  tinygrad/ops.py:95
                  │  │  └─ 0.425 run_ast  tinygrad/ops.py:97
                  │  │     └─ 0.410 <listcomp>  tinygrad/ops.py:98
                  │  │        └─ 0.410 run_ast  tinygrad/ops.py:97
                  │  │           └─ 0.401 <listcomp>  tinygrad/ops.py:98
                  │  │              └─ 0.401 run_ast  tinygrad/ops.py:97
                  │  │                 └─ 0.360 <listcomp>  tinygrad/ops.py:98
                  │  │                    └─ 0.360 run_ast  tinygrad/ops.py:97
                  │  │                       └─ 0.330 <listcomp>  tinygrad/ops.py:98
                  │  │                          └─ 0.329 run_ast  tinygrad/ops.py:97
                  │  │                             └─ 0.281 <listcomp>  tinygrad/ops.py:98
                  │  │                                └─ 0.281 run_ast  tinygrad/ops.py:97
                  │  └─ 0.283 Linearizer.simplify_merge_adjacent  tinygrad/codegen/kernel.py:249
                  └─ 0.325 Linearizer.hand_coded_optimizations  tinygrad/codegen/kernel.py:471
```

It looks like it wastes a ton of time in dlopen()

Reply:
![image](https://github.com/tinygrad/tinygrad/assets/83587632/552e3a86-2ea2-46ac-ac48-37d8311ed6cf)
Reproduced with just adding two tensors. It looks like for some reason dlopen in python is horribly slow, thus high kernel dispatch overhead, thus it takes ages when there is a lot of small computations (like in fuzzing stuff)

I can just disable it for clang, but that doesn't feel like a right fix

Reply:
Searched for how other fuzzing is done in tinygrad - it is in separate `tests/external` directory that isn't being ran by CI, should i just move my fuzzing here?

Reply:
Moved both fuzzing and benchmark to external

Reply:
symbolic and shapetracker fuzzer runs under a different worker and those are device agnostic

for testing multitensor allreduce, i think we want at least one interpreted backend and one compiled backend. fine to skip for clang if it's the only slow one due to writing to files

Reply:
Done